### PR TITLE
virt-launcher: log the domain XML upon launching

### PR DESCRIPTION
**Pull request by enp0s3**
Dump the domain XML of the VM to virt-launcher log. The XML dump will be encoded with Base64
since the security measurements of the JSON logger makes the XML unreadable due to special character escaping.

Signed-off-by: Igor Bezukh <ibezukh@redhat.com>

**What this PR does / why we need it**:
To ease the debugging of issues related to VM launch failures. 

**Which issue(s) this PR fixes** 
https://bugzilla.redhat.com/show_bug.cgi?id=1848403

```release-note
None
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 1]
```diff
@@ -21,6 +23,8 @@ import (
 	"kubevirt.io/client-go/log"
 	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
 	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/cli"
+
+	"github.com/go-xmlfmt/xmlfmt"
```
Comment (danielBelenky) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:27]: I don't think that you need this library.
You can use something like:

```
	dom := &libvirtxml.Domain{}
	_ = xml.Unmarshal([]byte(domain), &dom)
	buf, _ := xml.MarshalIndent(dom, "", "\t")
	fmt.Println(string(buf))
```

and then we use the official xml library and libvirt's xml library.

Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @danielBelenky Thanks for the advice! I've updated the PR

File: examples/vmi-ephemeral.yaml [hunk 2]
```diff
@@ -20,5 +20,5 @@ spec:
   terminationGracePeriodSeconds: 0
   volumes:
   - containerDisk:
-      image: registry:5000/kubevirt/cirros-container-disk-demo:devel
+      image: kubevirt/cirros-container-disk-demo:latest
```
Comment (danielBelenky) [inline examples/vmi-ephemeral.yaml:23]: Please remove unrelated changes from the PR

Comment (enp0s3) [reply examples/vmi-ephemeral.yaml]: @danielBelenky Leaked by mistake. Done

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 3]
```diff
@@ -408,3 +412,12 @@ func SetupLibvirt() error {
 
 	return nil
 }
+
+func convertToLibvirtXML(domspec string) string {
+
+	dom := &libvirtxml.Domain{}
+	_ = xml.Unmarshal([]byte(domspec), &dom)
```
Comment (qinqon) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:419]: Can we at least Print the error ?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 4]
```diff
@@ -408,3 +412,12 @@ func SetupLibvirt() error {
...
+	buf, _ := xml.MarshalIndent(dom, "", "\t")
```
Comment (qinqon) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:420]: ditto

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 5]
```diff
@@ -92,7 +93,10 @@ func ConvReason(status libvirt.DomainState, reason int) api.StateChangeReason {
 }
 
 func SetDomainSpecStr(virConn cli.Connection, vmi *v1.VirtualMachineInstance, wantedSpec string) (cli.VirDomain, error) {
-	log.Log.Object(vmi).V(3).With("xml", wantedSpec).Info("Domain XML generated.")
+
+	fmt.Println("Domain XML generated.")
+	fmt.Println(convertToLibvirtXML(wantedSpec))
```
Comment (qinqon) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:98]: Don't we want to use log ?

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:98]: +1 would that mean that we just drop a xml in the middle of the log stream?

Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon @rmohr the logger we use is based on JSON logger and it is set with escapeHTMLLtrue which means the >< & characters will be escaped. I don't see an option to override this value when calliing to log.Log(..) @rmohr the log stream looks coherent with this print, so per the validations I made on local setup.

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 4 above)
Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon Done

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 3 above)
Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon Done

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 5 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @enp0s3 can you paste a chunk of how log looks like ?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 6]
```diff
@@ -408,3 +416,18 @@ func SetupLibvirt() error {
 
 	return nil
 }
+
+func convertToLibvirtXML(domspec string) (string, error) {
+
+	dom := &libvirtxml.Domain{}
+	err := xml.Unmarshal([]byte(domspec), &dom)
+	if err != nil {
+		return "", err
```
Comment (qinqon) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:425]: please do a `fmt.Errorf("failed unmarshaling libvirt xml: %v", err)` so it's easier to find where it's comming from

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 7]
```diff
@@ -408,3 +416,18 @@ func SetupLibvirt() error {
...
+	}
+	buf, err := xml.MarshalIndent(dom, "", "\t")
+	if err != nil {
+		return "", err
```
Comment (qinqon) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:429]: ditto

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 5 above)
Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon Sure. [link](https://raw.githubusercontent.com/enp0s3/tools/master/xml_log.txt)

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 7 above)
Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon Done

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 8]
```diff
@@ -92,7 +93,14 @@ func ConvReason(status libvirt.DomainState, reason int) api.StateChangeReason {
 }
 
 func SetDomainSpecStr(virConn cli.Connection, vmi *v1.VirtualMachineInstance, wantedSpec string) (cli.VirDomain, error) {
-	log.Log.Object(vmi).V(3).With("xml", wantedSpec).Info("Domain XML generated.")
+
+	domxml, err := convertToLibvirtXML(wantedSpec)
+	if err == nil {
+		fmt.Println("Domain XML generated.")
+		fmt.Println(domxml)
```
Comment (qinqon) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:100]: Maybe we cannot use structured logs from k8s but we can still use golang log package to get at least the timestamps

https://golang.org/pkg/log/

Also add the vmi name to it so it's easier to track.

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 5 above)
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: What is the issue with the escapes?

Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @rmohr It's not human readable, for example, it will look like this:
```
"\u003cdomain type=\"kvm\"\u003e\n\t\u003cname\u003edefault_vmi-ephemeral\u003c/name\u003e\n\t\u003cmetadata\u003e\u003ckubevirt xmlns=\"http://kubevirt.io\"\u003e\u003cuid\u003e1a978d61-a6a0-4f37-a460-ec3369c813d0\u003c/uid\u003e\u003cgraceperiod\u003e\u003cdeletionGracePeriodSeconds\u003e0\u003c/deletionGracePeriodSeconds\u003e\u003c/graceperiod\u003e\u003c/kubevirt\u003e\u003c/metadata\u003e\n\t\u003cmemory unit=\"b\"\u003e64000000\u003c/memory\u003e\n\t\u003cvcpu placement=\"static\"\u003e1\u003c/vcpu\u003e\n\t\u003ciothreads\u003e1\u003c/iothreads\u003e\n\t\u003csysinfo type=\"smbios\"\u003e\n\t\t\u003cbios\u003e\u003c/bios\u003e\n\t\t\u003csystem\u003e\n\t\t\t\u003centry name=\"uuid\"\u003e44c70f30-2bb3-4b54-8e9d-22046d72df3f\u003c/entry\u003e\n\t\t\t\u003centry name=\"manufacturer\"\u003eKubeVirt\u003c/entry\u003e\n\t\t\t\u003centry name=\"family\"\u003eKubeVirt\u003c/entry\u003e\n\t\t\t\u003centry name=\"product\"\u003eNone\u003c/entry\u003e\n\t\t\t\u003centry name=\"sku\"\u003e\u003c/entry\u003e\n\t\t\t\u003centry name=\"version\"\u003e\u003c/entry\u003e\n\t\t\u003c/system\u003e\n\t\t\u003cbaseBoard\u003e\u003c/baseBoard\u003e\n\t\t\u003cchassis\u003e\u003c/chassis\u003e\n\t\u003c/sysinfo\u003e\n\t\u003cos\u003e\n\t\t\u003ctype arch=\"x86_64\" machine=\"q35\"\u003ehvm\u003c/type\u003e\n\t\t\u003csmbios mode=\"sysinfo\"\u003e\u003c/smbios\u003e\n\t\u003c/os\u003e\n\t\u003cfeatures\u003e\n\t\t\u003cacpi\u003e\u003c/acpi\u003e\n\t\u003c/features\u003e\n\t\u003ccpu mode=\"host-model\"\u003e\n\t\t\u003ctopology sockets=\"1\" cores=\"1\" threads=\"1\"\u003e\u003c/topology\u003e\n\t\u003c/cpu\u003e\n\t\u003cdevices\u003e\n\t\t\u003cdisk type=\"file\" device=\"disk\"\u003e\n\t\t\t\u003cdriver name=\"qemu\" type=\"qcow2\" cache=\"none\"\u003e\u003c/driver\u003e\n\t\t\t\u003csource file=\"/var/run/kubevirt-ephemeral-disks/disk-data/containerdisk/disk.qcow2\"\u003e\u003c/source\u003e\n\t\t\t\u003cbackingStore type=\"file\"\u003e\n\t\t\t\t\u003cformat type=\"qcow2\"\u003e\u003c/format\u003e\n\t\t\t\t\u003csource file=\"/var/run/kubevirt/container-disks/disk_0.img\"\u003e\u003c/source\u003e\n\t\t\t\u003c/backingStore\u003e\n\t\t\t\u003ctarget dev=\"vda\" bus=\"virtio\"\u003e\u003c/target\u003e\n\t\t\t\u003calias name=\"ua-containerdisk\"\u003e\u003c/alias\u003e\n\t\t\u003c/disk\u003e\n\t\t\u003ccontroller type=\"usb\" index=\"0\" model=\"none\"\u003e\u003c/controller\u003e\n\t\t\u003ccontroller type=\"virtio-serial\" index=\"0\"\u003e\u003c/controller\u003e\n\t\t\u003cinterface type=\"bridge\"\u003e\n\t\t\t\u003cmac address=\"0e:7d:82:83:7f:3a\"\u003e\u003c/mac\u003e\n\t\t\t\u003csource bridge=\"k6t-eth0\"\u003e\u003c/source\u003e\n\t\t\t\u003cmodel type=\"virtio\"\u003e\u003c/model\u003e\n\t\t\t\u003cmtu size=\"1440\"\u003e\u003c/mtu\u003e\n\t\t\t\u003calias name=\"ua-default\"\u003e\u003c/alias\u003e\n\t\t\u003c/interface\u003e\n\t\t\u003cserial type=\"unix\"\u003e\n\t\t\t\u003csource mode=\"bind\" path=\"/var/run/kubevirt-private/1a978d61-a6a0-4f37-a460-ec3369c813d0/virt-serial0\"\u003e\u003c/source\u003e\n\t\t\t\u003ctarget port=\"0\"\u003e\u003c/target\u003e\n\t\t\u003c/serial\u003e\n\t\t\u003cconsole type=\"pty\"\u003e\n\t\t\t\u003ctarget type=\"serial\" port=\"0\"\u003e\u003c/target\u003e\n\t\t\u003c/console\u003e\n\t\t\u003cchannel type=\"unix\"\u003e\n\t\t\t\u003ctarget type=\"virtio\" name=\"org.qemu.guest_agent.0\"\u003e\u003c/target\u003e\n\t\t\u003c/channel\u003e\n\t\t\u003cgraphics type=\"vnc\"\u003e\n\t\t\t\u003clisten type=\"socket\" socket=\"/var/run/kubevirt-private/1a978d61-a6a0-4f37-a460-ec3369c813d0/virt-vnc\"\u003e\u003c/listen\u003e\n\t\t\u003c/graphics\u003e\n\t\t\u003cvideo\u003e\n\t\t\t\u003cmodel type=\"vga\" heads=\"1\" vram=\"16384\"\u003e\u003c/model\u003e\n\t\t\u003c/video\u003e\n\t\t\u003cmemballoon model=\"virtio\"\u003e\n\t\t\t\u003cstats period=\"10\"\u003e\u003c/stats\u003e\n\t\t\u003c/memballoon\u003e\n\t\u003c/devices\u003e\n\u003c/domain\u003e\n","pos":"libvirt_helper.go:100","timestamp":"2020-08-05T15:49:46.732033Z"}
```

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go (hunk 8 above)
Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon k8s logger uses golang log and overrides the IO writer. So logging will still occur in JSON format with escapes. Looks like Logrus logging package can be used instead.

Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @qinqon @rmohr Why can't I just print it with fmt.Println?

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: Perhaps all we need is something like [1]?
It's very small and seems to do the job. Then we could simply log the output.
What do you think?

[1] https://github.com/go-xmlfmt/xmlfmt/blob/master/xmlfmt.go

Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @vladikr I've suggested that package in the initial PR phase. Later on together with @danielBelenky we decided to use a more maintained package libvirt-go-xml.

**Comment by enp0s3** [comment]
> I've only briefly looked at this pr. I think the idea is fine. Can someone verify there's nothing in the domain xml that exposes any secret/password data before merging this. thanks!

Just by looking at https://libvirt.org/formatdomain.html and searching for possible exposed credentials, I didn't find sensitive elements that may be shown in the XML. There is the **_auth_** but it hides the password. Besides that, If you can point me to an additional source of information it will be great.

**Comment by davidvossel** [comment]
> @davidvossel Just by looking at https://libvirt.org/formatdomain.html and searching for possible exposed credentials, I didn't find sensitive elements that may be shown in the XML. There is the auth but it hides the password. Besides that, If you can point me to an additional source of information it will be great.

nope, that sounds good enough to me. I think we're set on that front.

For outputting the log, we just talked about it in person, but I'll post my thoughts here for a wider audience.

It seems to me that trying to make the domain xml human readable in a json structured log is super challenging. I'm not sure we'll ever get something that is truly human readable where someone won't copy the xml log lines and at least run them through an xml viewer of some sort to make the indention and everything look right.

If we're spinning our wheels here, my vote is to simply dump the domainxml as a base64 encoded blob to the log. If someone wants to inspect it, then they can copy that blob and base64 decode it. The big thing is the information is there one way or another.

**Comment by enp0s3** [comment]
@davidvossel @rmohr @vladikr @qinqon 

As per David's suggestion, the log will look like this:

`{"component":"virt-launcher","kind":"","level":"info","msg":"Domain XML generated. XML dump encoded with Base64: PGRvbWFpbiB0eXBlPSJrdm0iPgoJPG5hbWU+ZGVmYXVsdF92bWktZXBoZW1lcmFsPC9uYW1lPgoJPG1ldGFkYXRhPjxrdWJldmlydCB4bWxucz0iaHR0cDovL2t1YmV2aXJ0LmlvIj48dWlkPjE0NDI3NTUzLTY0YzUtNGU5Ny05YmI5LTJlYjU4MTgwMmQ4YzwvdWlkPjxncmFjZXBlcmlvZD48ZGVsZXRpb25HcmFjZVBlcmlvZFNlY29uZHM+MDwvZGVsZXRpb25HcmFjZVBlcmlvZFNlY29uZHM+PC9ncmFjZXBlcmlvZD48L2t1YmV2aXJ0PjwvbWV0YWRhdGE+Cgk8bWVtb3J5IHVuaXQ9ImIiPjY0MDAwMDAwPC9tZW1vcnk+Cgk8dmNwdSBwbGFjZW1lbnQ9InN0YXRpYyI+MTwvdmNwdT4KCTxpb3RocmVhZHM+MTwvaW90aHJlYWRzPgoJPHN5c2luZm8gdHlwZT0ic21iaW9zIj4KCQk8Ymlvcz48L2Jpb3M+CgkJPHN5c3RlbT4KCQkJPGVudHJ5IG5hbWU9InV1aWQiPmYzZGRmNmNmLTA1MzgtNDk3ZS1iYWFlLTRiZjJkMjVjZjFlZDwvZW50cnk+CgkJCTxlbnRyeSBuYW1lPSJtYW51ZmFjdHVyZXIiPkt1YmVWaXJ0PC9lbnRyeT4KCQkJPGVudHJ5IG5hbWU9ImZhbWlseSI+S3ViZVZpcnQ8L2VudHJ5PgoJCQk8ZW50cnkgbmFtZT0icHJvZHVjdCI+Tm9uZTwvZW50cnk+CgkJCTxlbnRyeSBuYW1lPSJza3UiPjwvZW50cnk+CgkJCTxlbnRyeSBuYW1lPSJ2ZXJzaW9uIj48L2VudHJ5PgoJCTwvc3lzdGVtPgoJCTxiYXNlQm9hcmQ+PC9iYXNlQm9hcmQ+CgkJPGNoYXNzaXM+PC9jaGFzc2lzPgoJPC9zeXNpbmZvPgoJPG9zPgoJCTx0eXBlIGFyY2g9Ing4Nl82NCIgbWFjaGluZT0icTM1Ij5odm08L3R5cGU+CgkJPHNtYmlvcyBtb2RlPSJzeXNpbmZvIj48L3NtYmlvcz4KCTwvb3M+Cgk8ZmVhdHVyZXM+CgkJPGFjcGk+PC9hY3BpPgoJPC9mZWF0dXJlcz4KCTxjcHUgbW9kZT0iaG9zdC1tb2RlbCI+CgkJPHRvcG9sb2d5IHNvY2tldHM9IjEiIGNvcmVzPSIxIiB0aHJlYWRzPSIxIj48L3RvcG9sb2d5PgoJPC9jcHU+Cgk8ZGV2aWNlcz4KCQk8ZGlzayB0eXBlPSJmaWxlIiBkZXZpY2U9ImRpc2siPgoJCQk8ZHJpdmVyIG5hbWU9InFlbXUiIHR5cGU9InFjb3cyIiBjYWNoZT0ibm9uZSI+PC9kcml2ZXI+CgkJCTxzb3VyY2UgZmlsZT0iL3Zhci9ydW4va3ViZXZpcnQtZXBoZW1lcmFsLWRpc2tzL2Rpc2stZGF0YS9jb250YWluZXJkaXNrL2Rpc2sucWNvdzIiPjwvc291cmNlPgoJCQk8YmFja2luZ1N0b3JlIHR5cGU9ImZpbGUiPgoJCQkJPGZvcm1hdCB0eXBlPSJxY293MiI+PC9mb3JtYXQ+CgkJCQk8c291cmNlIGZpbGU9Ii92YXIvcnVuL2t1YmV2aXJ0L2NvbnRhaW5lci1kaXNrcy9kaXNrXzAuaW1nIj48L3NvdXJjZT4KCQkJPC9iYWNraW5nU3RvcmU+CgkJCTx0YXJnZXQgZGV2PSJ2ZGEiIGJ1cz0idmlydGlvIj48L3RhcmdldD4KCQkJPGFsaWFzIG5hbWU9InVhLWNvbnRhaW5lcmRpc2siPjwvYWxpYXM+CgkJPC9kaXNrPgoJCTxjb250cm9sbGVyIHR5cGU9InVzYiIgaW5kZXg9IjAiIG1vZGVsPSJub25lIj48L2NvbnRyb2xsZXI+CgkJPGNvbnRyb2xsZXIgdHlwZT0idmlydGlvLXNlcmlhbCIgaW5kZXg9IjAiPjwvY29udHJvbGxlcj4KCQk8aW50ZXJmYWNlIHR5cGU9ImJyaWRnZSI+CgkJCTxtYWMgYWRkcmVzcz0iZDY6ZGI6ZTQ6MjI6ZDQ6Y2UiPjwvbWFjPgoJCQk8c291cmNlIGJyaWRnZT0iazZ0LWV0aDAiPjwvc291cmNlPgoJCQk8bW9kZWwgdHlwZT0idmlydGlvIj48L21vZGVsPgoJCQk8bXR1IHNpemU9IjE0NTAiPjwvbXR1PgoJCQk8YWxpYXMgbmFtZT0idWEtZGVmYXVsdCI+PC9hbGlhcz4KCQk8L2ludGVyZmFjZT4KCQk8c2VyaWFsIHR5cGU9InVuaXgiPgoJCQk8c291cmNlIG1vZGU9ImJpbmQiIHBhdGg9Ii92YXIvcnVuL2t1YmV2aXJ0LXByaXZhdGUvMTQ0Mjc1NTMtNjRjNS00ZTk3LTliYjktMmViNTgxODAyZDhjL3ZpcnQtc2VyaWFsMCI+PC9zb3VyY2U+CgkJCTx0YXJnZXQgcG9ydD0iMCI+PC90YXJnZXQ+CgkJPC9zZXJpYWw+CgkJPGNvbnNvbGUgdHlwZT0icHR5Ij4KCQkJPHRhcmdldCB0eXBlPSJzZXJpYWwiIHBvcnQ9IjAiPjwvdGFyZ2V0PgoJCTwvY29uc29sZT4KCQk8Y2hhbm5lbCB0eXBlPSJ1bml4Ij4KCQkJPHRhcmdldCB0eXBlPSJ2aXJ0aW8iIG5hbWU9Im9yZy5xZW11Lmd1ZXN0X2FnZW50LjAiPjwvdGFyZ2V0PgoJCTwvY2hhbm5lbD4KCQk8Z3JhcGhpY3MgdHlwZT0idm5jIj4KCQkJPGxpc3RlbiB0eXBlPSJzb2NrZXQiIHNvY2tldD0iL3Zhci9ydW4va3ViZXZpcnQtcHJpdmF0ZS8xNDQyNzU1My02NGM1LTRlOTctOWJiOS0yZWI1ODE4MDJkOGMvdmlydC12bmMiPjwvbGlzdGVuPgoJCTwvZ3JhcGhpY3M+CgkJPHZpZGVvPgoJCQk8bW9kZWwgdHlwZT0idmdhIiBoZWFkcz0iMSIgdnJhbT0iMTYzODQiPjwvbW9kZWw+CgkJPC92aWRlbz4KCQk8bWVtYmFsbG9vbiBtb2RlbD0idmlydGlvIj4KCQkJPHN0YXRzIHBlcmlvZD0iMTAiPjwvc3RhdHM+CgkJPC9tZW1iYWxsb29uPgoJPC9kZXZpY2VzPgo8L2RvbWFpbj4=","name":"vmi-ephemeral","namespace":"default","pos":"libvirt_helper.go:100","timestamp":"2020-08-13T12:03:25.605543Z","uid":"14427553-64c5-4e97-9bb9-2eb581802d8c"}
{"component":"virt-launcher","kind":"","level":"info","msg":"Domain defined.","name":"vmi-ephemeral","namespace":"default","pos":"manager.go:1213","timestamp":"2020-08-13T12:03:25.959699Z","uid":"14427553-64c5-4e97-9bb9-2eb581802d8c"}
`

Is that a suitable solution?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go [hunk 9]
```diff
@@ -408,3 +415,17 @@ func SetupLibvirt() error {
 
 	return nil
 }
+
+func convertToLibvirtXML(domspec string) ([]byte, error) {
```
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:419]: can we get a unit test for this

Comment (enp0s3) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: @davidvossel I've removed this function since we actually don't need it anymore. Also, there is no need in libvirt-go-xml package for proper XML format. We already have a proper domain spec from the libvirt-go API. I only changed the way XML marshals the domain spec so it will be more readable, and encoded it while calling the logger.

**Comment by enp0s3** [comment]
@davidvossel @stu-gott If we confirm that Base64 encoded approach then the changes are minor.

**Comment by danielBelenky** [comment]
/lgtm
/approve

PR was updated just to regenerate a token for travis.

**Comment by rmohr** [comment]
We had to migrate a job from Travis to prow which required manual merge. Since this messes up the merge pool, helping prow a little with merging trivial PRs.
